File: python/src/bleepstore/storage/local.py
"""Local filesystem storage backend for BleepStore.

Implements the StorageBackend protocol using the local filesystem.
Objects are stored under ``{root}/{bucket}/{key}``. Multipart parts
are stored under ``{root}/.parts/{upload_id}/{part_number}``.

Crash-only design:
    - Atomic writes via temp-fsync-rename pattern.
    - Never acknowledge before data is fsync'd to disk.
    - Startup cleans orphan temp files in ``.tmp/`` subdirectories.
"""

import hashlib
import logging
import os
import uuid
from collections.abc import AsyncIterator
from pathlib import Path

logger = logging.getLogger(__name__)

# Streaming chunk size: 64 KB
_CHUNK_SIZE = 64 * 1024


class LocalStorageBackend:
    """Storage backend that persists objects on the local filesystem.

    Objects are stored under ``{root}/{bucket}/{key}``. Multipart parts
    are stored under ``{root}/.parts/{upload_id}/{part_number}``.

    Attributes:
        root: The root directory for all stored data.
    """

    def __init__(self, root: str | Path) -> None:
        """Initialize the local storage backend.

        Args:
            root: Root directory path for object storage.
        """
        self.root = Path(root)

    def _object_path(self, bucket: str, key: str) -> Path:
        """Return the filesystem path for a stored object.

        Args:
            bucket: The bucket name.
            key: The object key.

        Returns:
            The absolute path where this object is stored.
        """
        return self.root / bucket / key

    async def init(self) -> None:
        """Create the root directory and clean up orphan temp files.

        Crash-only design: every startup is a recovery. Remove any
        leftover ``.tmp.*`` files from interrupted writes.
        """
        self.root.mkdir(parents=True, exist_ok=True)

        # Clean orphan temp files from previous crashes
        self._clean_temp_files()

        logger.info("Local storage backend initialized at %s", self.root)

    def _clean_temp_files(self) -> None:
        """Remove orphan temp files left by interrupted atomic writes."""
        count = 0
        for dirpath, _dirnames, filenames in os.walk(self.root):
            for fname in filenames:
                if ".tmp." in fname:
                    try:
                        os.unlink(os.path.join(dirpath, fname))
                        count += 1
                    except OSError:
                        pass
        if count > 0:
            logger.info("Cleaned %d orphan temp files on startup", count)

    async def close(self) -> None:
        """No-op for local filesystem backend."""
        pass

    async def put(self, bucket: str, key: str, data: bytes) -> str:
        """Store an object's bytes on the local filesystem.

        Uses the atomic temp-fsync-rename pattern for crash safety.
        Never acknowledges before data is committed to disk.

        Args:
            bucket: The bucket name.
            key: The object key.
            data: The raw bytes to store.

        Returns:
            The hex-encoded MD5 of the stored data.
        """
        path = self._object_path(bucket, key)
        path.parent.mkdir(parents=True, exist_ok=True)

        # Compute MD5
        md5 = hashlib.md5(data).hexdigest()

        # Atomic write: temp file -> fsync -> rename
        tmp = path.with_name(f"{path.name}.tmp.{uuid.uuid4().hex[:8]}")
        try:
            fd = os.open(str(tmp), os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
            try:
                os.write(fd, data)
                os.fsync(fd)
            finally:
                os.close(fd)
            tmp.rename(path)
        except Exception:
            # Clean up temp file on failure
            try:
                tmp.unlink(missing_ok=True)
            except OSError:
                pass
            raise

        return md5

    async def get(self, bucket: str, key: str) -> bytes:
        """Retrieve an object's bytes from the local filesystem.

        Args:
            bucket: The bucket name.
            key: The object key.

        Returns:
            The raw bytes of the object.

        Raises:
            FileNotFoundError: If the object does not exist on disk.
        """
        path = self._object_path(bucket, key)
        return path.read_bytes()

    async def get_stream(
        self, bucket: str, key: str, offset: int = 0, length: int | None = None
    ) -> AsyncIterator[bytes]:
        """Retrieve an object as an async byte stream.

        Yields 64 KB chunks from the specified offset up to ``length``
        bytes (or end of file if ``length`` is None).

        Args:
            bucket: The bucket name.
            key: The object key.
            offset: Byte offset to start reading from.
            length: Number of bytes to read, or None for all remaining.

        Yields:
            Chunks of bytes from the object.
        """
        path = self._object_path(bucket, key)
        remaining = length

        with open(path, "rb") as f:
            if offset > 0:
                f.seek(offset)

            while True:
                if remaining is not None:
                    to_read = min(_CHUNK_SIZE, remaining)
                    if to_read <= 0:
                        break
                else:
                    to_read = _CHUNK_SIZE

                chunk = f.read(to_read)
                if not chunk:
                    break

                yield chunk

                if remaining is not None:
                    remaining -= len(chunk)

    async def delete(self, bucket: str, key: str) -> None:
        """Delete an object from the local filesystem.

        Silently ignores missing files (idempotent). Cleans up empty
        parent directories up to the bucket directory.

        Args:
            bucket: The bucket name.
            key: The object key.
        """
        path = self._object_path(bucket, key)

        try:
            path.unlink()
        except FileNotFoundError:
            return

        # Clean up empty parent directories (up to bucket dir)
        bucket_dir = self.root / bucket
        parent = path.parent
        while parent != bucket_dir and parent != self.root:
            try:
                parent.rmdir()  # Only removes empty dirs
            except OSError:
                break
            parent = parent.parent

    async def exists(self, bucket: str, key: str) -> bool:
        """Check if an object exists on the local filesystem.

        Args:
            bucket: The bucket name.
            key: The object key.

        Returns:
            True if the file exists.
        """
        return self._object_path(bucket, key).is_file()

    async def put_part(
        self, bucket: str, key: str, upload_id: str, part_number: int, data: bytes
    ) -> str:
        """Store a multipart upload part on the local filesystem.

        Parts are stored under ``{root}/.parts/{upload_id}/{part_number}``.
        Uses atomic temp-fsync-rename for crash safety.

        Args:
            bucket: The bucket name.
            key: The object key.
            upload_id: The multipart upload identifier.
            part_number: The sequential part number.
            data: The raw bytes of this part.

        Returns:
            The hex-encoded MD5 ETag of the part.
        """
        part_dir = self.root / ".parts" / upload_id
        part_dir.mkdir(parents=True, exist_ok=True)
        part_path = part_dir / str(part_number)

        md5 = hashlib.md5(data).hexdigest()

        # Atomic write
        tmp = part_path.with_name(f"{part_path.name}.tmp.{uuid.uuid4().hex[:8]}")
        try:
            fd = os.open(str(tmp), os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
            try:
                os.write(fd, data)
                os.fsync(fd)
            finally:
                os.close(fd)
            tmp.rename(part_path)
        except Exception:
            try:
                tmp.unlink(missing_ok=True)
            except OSError:
                pass
            raise

        return md5

    async def assemble_parts(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_numbers: list[int],
    ) -> str:
        """Assemble uploaded parts into a final object on disk.

        Reads each part sequentially, writes concatenated to a temp file,
        then atomically renames to the final object path. Cleans up part
        files after successful assembly.

        Args:
            bucket: The bucket name.
            key: The object key.
            upload_id: The multipart upload identifier.
            part_numbers: Ordered list of part numbers to assemble.

        Returns:
            The hex-encoded MD5 ETag of the assembled object.
        """
        dest = self._object_path(bucket, key)
        dest.parent.mkdir(parents=True, exist_ok=True)

        tmp = dest.with_name(f"{dest.name}.tmp.{uuid.uuid4().hex[:8]}")
        md5 = hashlib.md5()

        try:
            fd = os.open(str(tmp), os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
            try:
                for pn in part_numbers:
                    part_path = self.root / ".parts" / upload_id / str(pn)
                    data = part_path.read_bytes()
                    os.write(fd, data)
                    md5.update(data)
                os.fsync(fd)
            finally:
                os.close(fd)
            tmp.rename(dest)
        except Exception:
            try:
                tmp.unlink(missing_ok=True)
            except OSError:
                pass
            raise

        return md5.hexdigest()

    async def delete_parts(self, bucket: str, key: str, upload_id: str) -> None:
        """Delete all stored parts for a multipart upload.

        Removes the upload directory and all part files within it.

        Args:
            bucket: The bucket name.
            key: The object key.
            upload_id: The multipart upload identifier.
        """
        parts_dir = self.root / ".parts" / upload_id
        if not parts_dir.exists():
            return

        # Remove all part files
        for child in parts_dir.iterdir():
            try:
                child.unlink()
            except OSError:
                pass

        # Remove the directory
        try:
            parts_dir.rmdir()
        except OSError:
            pass

    async def copy_object(
        self,
        src_bucket: str,
        src_key: str,
        dst_bucket: str,
        dst_key: str,
    ) -> str:
        """Copy an object from one location to another.

        Reads source bytes and writes to destination atomically.

        Args:
            src_bucket: The source bucket name.
            src_key: The source object key.
            dst_bucket: The destination bucket name.
            dst_key: The destination object key.

        Returns:
            The hex-encoded MD5 ETag of the copied object.
        """
        data = await self.get(src_bucket, src_key)
        return await self.put(dst_bucket, dst_key, data)

File: golang/internal/storage/local.go
package storage

import (
	"context"
	"crypto/md5"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/bleepstore/bleepstore/internal/uid"
)

// LocalBackend implements the StorageBackend interface using the local
// filesystem. Objects are stored as files within a configurable root directory,
// organized by bucket and key path.
type LocalBackend struct {
	// RootDir is the base directory under which all bucket and object data
	// is stored.
	RootDir string
}

// NewLocalBackend creates a new LocalBackend rooted at the given directory.
// It creates the root directory and the temp directory if they do not exist.
func NewLocalBackend(rootDir string) (*LocalBackend, error) {
	if err := os.MkdirAll(rootDir, 0o755); err != nil {
		return nil, fmt.Errorf("creating storage root directory %q: %w", rootDir, err)
	}
	// Create the .tmp directory for atomic writes.
	tmpDir := filepath.Join(rootDir, ".tmp")
	if err := os.MkdirAll(tmpDir, 0o755); err != nil {
		return nil, fmt.Errorf("creating temp directory %q: %w", tmpDir, err)
	}
	return &LocalBackend{RootDir: rootDir}, nil
}

// CleanTempFiles removes all files in the .tmp directory. This is called on
// startup as part of crash-only recovery. Any temp files left behind indicate
// incomplete writes from a previous crash.
func (b *LocalBackend) CleanTempFiles() error {
	tmpDir := filepath.Join(b.RootDir, ".tmp")
	entries, err := os.ReadDir(tmpDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("reading temp directory: %w", err)
	}
	for _, entry := range entries {
		if !entry.IsDir() {
			os.Remove(filepath.Join(tmpDir, entry.Name()))
		}
	}
	return nil
}

// objectPath returns the full filesystem path for an object.
func (b *LocalBackend) objectPath(bucket, key string) string {
	return filepath.Join(b.RootDir, bucket, key)
}

// tempPath returns a unique temporary file path in the .tmp directory.
func (b *LocalBackend) tempPath() string {
	return filepath.Join(b.RootDir, ".tmp", "tmp-"+uid.New())
}

// PutObject writes object data to a file on the local filesystem using the
// crash-only atomic write pattern: write to temp file, fsync, rename.
// Returns the number of bytes written and the ETag (MD5 hex digest).
func (b *LocalBackend) PutObject(ctx context.Context, bucket, key string, reader io.Reader, size int64) (int64, string, error) {
	objPath := b.objectPath(bucket, key)

	// Ensure parent directories exist.
	if err := os.MkdirAll(filepath.Dir(objPath), 0o755); err != nil {
		return 0, "", fmt.Errorf("creating parent directories for %q/%q: %w", bucket, key, err)
	}

	// Write to a temp file, computing MD5 along the way.
	tmpPath := b.tempPath()
	tmpFile, err := os.Create(tmpPath)
	if err != nil {
		return 0, "", fmt.Errorf("creating temp file: %w", err)
	}

	// Hash while writing via TeeReader.
	h := md5.New()
	tee := io.TeeReader(reader, h)

	bytesWritten, err := io.Copy(tmpFile, tee)
	if err != nil {
		tmpFile.Close()
		os.Remove(tmpPath)
		return 0, "", fmt.Errorf("writing object data: %w", err)
	}

	// Fsync before rename to guarantee durability.
	if err := tmpFile.Sync(); err != nil {
		tmpFile.Close()
		os.Remove(tmpPath)
		return 0, "", fmt.Errorf("syncing temp file: %w", err)
	}

	if err := tmpFile.Close(); err != nil {
		os.Remove(tmpPath)
		return 0, "", fmt.Errorf("closing temp file: %w", err)
	}

	// Atomic rename: temp -> final path.
	if err := os.Rename(tmpPath, objPath); err != nil {
		os.Remove(tmpPath)
		return 0, "", fmt.Errorf("renaming temp file to final path: %w", err)
	}

	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))
	return bytesWritten, etag, nil
}

// GetObject opens the object file for reading. Returns the file as a
// ReadCloser, the file size, and an empty ETag (metadata store holds the ETag).
// The caller is responsible for closing the returned ReadCloser.
func (b *LocalBackend) GetObject(ctx context.Context, bucket, key string) (io.ReadCloser, int64, string, error) {
	objPath := b.objectPath(bucket, key)

	file, err := os.Open(objPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, 0, "", fmt.Errorf("object not found: %s/%s", bucket, key)
		}
		return nil, 0, "", fmt.Errorf("opening object file %q/%q: %w", bucket, key, err)
	}

	info, err := file.Stat()
	if err != nil {
		file.Close()
		return nil, 0, "", fmt.Errorf("stat object file %q/%q: %w", bucket, key, err)
	}

	return file, info.Size(), "", nil
}

// DeleteObject removes the object file from the local filesystem.
// Idempotent: deleting a non-existent file is not an error.
// Also cleans up empty parent directories up to the bucket root.
func (b *LocalBackend) DeleteObject(ctx context.Context, bucket, key string) error {
	objPath := b.objectPath(bucket, key)

	err := os.Remove(objPath)
	if err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("removing object file %q/%q: %w", bucket, key, err)
	}

	// Clean up empty parent directories up to the bucket root.
	bucketDir := filepath.Join(b.RootDir, bucket)
	dir := filepath.Dir(objPath)
	for dir != bucketDir && dir != b.RootDir {
		if err := os.Remove(dir); err != nil {
			// Directory not empty or other error: stop climbing.
			break
		}
		dir = filepath.Dir(dir)
	}

	return nil
}

// CopyObject copies an object file from source to destination on the local
// filesystem using the atomic write pattern. Returns the new ETag.
func (b *LocalBackend) CopyObject(ctx context.Context, srcBucket, srcKey, dstBucket, dstKey string) (string, error) {
	srcPath := b.objectPath(srcBucket, srcKey)

	srcFile, err := os.Open(srcPath)
	if err != nil {
		if os.IsNotExist(err) {
			return "", fmt.Errorf("source object not found: %s/%s", srcBucket, srcKey)
		}
		return "", fmt.Errorf("opening source object: %w", err)
	}
	defer srcFile.Close()

	info, err := srcFile.Stat()
	if err != nil {
		return "", fmt.Errorf("stat source object: %w", err)
	}

	_, etag, err := b.PutObject(ctx, dstBucket, dstKey, srcFile, info.Size())
	if err != nil {
		return "", fmt.Errorf("copying object data: %w", err)
	}

	return etag, nil
}

// PutPart writes a single multipart upload part to the local filesystem.
func (b *LocalBackend) PutPart(ctx context.Context, bucket, key, uploadID string, partNumber int, reader io.Reader, size int64) (string, error) {
	partDir := filepath.Join(b.RootDir, ".multipart", uploadID)
	if err := os.MkdirAll(partDir, 0o755); err != nil {
		return "", fmt.Errorf("creating part directory: %w", err)
	}

	partPath := filepath.Join(partDir, fmt.Sprintf("%05d", partNumber))

	// Write to temp file, compute MD5, atomic rename.
	tmpPath := b.tempPath()
	tmpFile, err := os.Create(tmpPath)
	if err != nil {
		return "", fmt.Errorf("creating temp file for part: %w", err)
	}

	h := md5.New()
	tee := io.TeeReader(reader, h)

	if _, err := io.Copy(tmpFile, tee); err != nil {
		tmpFile.Close()
		os.Remove(tmpPath)
		return "", fmt.Errorf("writing part data: %w", err)
	}

	if err := tmpFile.Sync(); err != nil {
		tmpFile.Close()
		os.Remove(tmpPath)
		return "", fmt.Errorf("syncing part file: %w", err)
	}

	if err := tmpFile.Close(); err != nil {
		os.Remove(tmpPath)
		return "", fmt.Errorf("closing part temp file: %w", err)
	}

	if err := os.Rename(tmpPath, partPath); err != nil {
		os.Remove(tmpPath)
		return "", fmt.Errorf("renaming part temp file: %w", err)
	}

	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))
	return etag, nil
}

// AssembleParts concatenates the specified parts into a single object file.
// Uses atomic write pattern. Returns the composite ETag.
func (b *LocalBackend) AssembleParts(ctx context.Context, bucket, key, uploadID string, partNumbers []int) (string, error) {
	objPath := b.objectPath(bucket, key)
	if err := os.MkdirAll(filepath.Dir(objPath), 0o755); err != nil {
		return "", fmt.Errorf("creating parent directories: %w", err)
	}

	partDir := filepath.Join(b.RootDir, ".multipart", uploadID)
	tmpPath := b.tempPath()
	tmpFile, err := os.Create(tmpPath)
	if err != nil {
		return "", fmt.Errorf("creating temp file for assembly: %w", err)
	}

	// Concatenate parts and compute composite ETag from individual part MD5s.
	compositeMD5 := md5.New()
	for _, pn := range partNumbers {
		partPath := filepath.Join(partDir, fmt.Sprintf("%05d", pn))
		partFile, err := os.Open(partPath)
		if err != nil {
			tmpFile.Close()
			os.Remove(tmpPath)
			return "", fmt.Errorf("opening part %d: %w", pn, err)
		}

		// Compute MD5 of this part while copying.
		partHash := md5.New()
		tee := io.TeeReader(partFile, partHash)
		if _, err := io.Copy(tmpFile, tee); err != nil {
			partFile.Close()
			tmpFile.Close()
			os.Remove(tmpPath)
			return "", fmt.Errorf("copying part %d: %w", pn, err)
		}
		partFile.Close()

		// Add part MD5 to composite hash.
		compositeMD5.Write(partHash.Sum(nil))
	}

	if err := tmpFile.Sync(); err != nil {
		tmpFile.Close()
		os.Remove(tmpPath)
		return "", fmt.Errorf("syncing assembled file: %w", err)
	}

	if err := tmpFile.Close(); err != nil {
		os.Remove(tmpPath)
		return "", fmt.Errorf("closing assembled temp file: %w", err)
	}

	if err := os.Rename(tmpPath, objPath); err != nil {
		os.Remove(tmpPath)
		return "", fmt.Errorf("renaming assembled file: %w", err)
	}

	// Composite ETag format: "md5-of-concatenated-part-md5s-N"
	etag := fmt.Sprintf(`"%x-%d"`, compositeMD5.Sum(nil), len(partNumbers))

	// Clean up part files.
	os.RemoveAll(partDir)

	return etag, nil
}

// DeleteParts removes all part files associated with the given multipart upload.
func (b *LocalBackend) DeleteParts(ctx context.Context, bucket, key, uploadID string) error {
	partDir := filepath.Join(b.RootDir, ".multipart", uploadID)
	err := os.RemoveAll(partDir)
	if err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("removing part directory %q: %w", partDir, err)
	}

	// Best-effort cleanup: remove .multipart dir if empty.
	multipartDir := filepath.Join(b.RootDir, ".multipart")
	os.Remove(multipartDir) // Fails silently if not empty.

	return nil
}

// DeleteUploadParts removes the parts directory for a specific multipart upload.
// This is used during startup reaping of expired uploads to clean up orphaned
// part files on disk.
func (b *LocalBackend) DeleteUploadParts(uploadID string) error {
	partDir := filepath.Join(b.RootDir, ".multipart", uploadID)
	err := os.RemoveAll(partDir)
	if err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("removing part directory %q: %w", partDir, err)
	}

	// Best-effort cleanup: remove .multipart dir if empty.
	multipartDir := filepath.Join(b.RootDir, ".multipart")
	os.Remove(multipartDir) // Fails silently if not empty.

	return nil
}

// CreateBucket creates a directory for the bucket under the root directory.
func (b *LocalBackend) CreateBucket(ctx context.Context, bucket string) error {
	bucketDir := filepath.Join(b.RootDir, bucket)
	if err := os.MkdirAll(bucketDir, 0o755); err != nil {
		return fmt.Errorf("creating bucket directory %q: %w", bucketDir, err)
	}
	return nil
}

// DeleteBucket removes the bucket directory from the local filesystem.
// The directory must be empty.
func (b *LocalBackend) DeleteBucket(ctx context.Context, bucket string) error {
	bucketDir := filepath.Join(b.RootDir, bucket)
	// os.Remove only removes empty directories, which is the desired behavior.
	err := os.Remove(bucketDir)
	if err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("removing bucket directory %q: %w", bucketDir, err)
	}
	return nil
}

// ObjectExists checks whether an object exists on the local filesystem.
func (b *LocalBackend) ObjectExists(ctx context.Context, bucket, key string) (bool, error) {
	objPath := filepath.Join(b.RootDir, bucket, key)
	info, err := os.Stat(objPath)
	if err == nil {
		// Make sure it's a file, not a directory.
		if info.IsDir() {
			return false, nil
		}
		return true, nil
	}
	if os.IsNotExist(err) {
		return false, nil
	}
	return false, fmt.Errorf("checking object existence %q/%q: %w", bucket, key, err)
}

// HealthCheck verifies that the local storage root directory is accessible.
func (b *LocalBackend) HealthCheck(ctx context.Context) error {
	_, err := os.Stat(b.RootDir)
	return err
}

// cleanEmptyParents removes empty directories starting from dir up to (but not
// including) stopAt. This is useful for cleaning up after object deletion when
// keys contain "/" separators that create subdirectories.
func cleanEmptyParents(dir, stopAt string) {
	// Normalize paths for comparison.
	dir = filepath.Clean(dir)
	stopAt = filepath.Clean(stopAt)

	for dir != stopAt && strings.HasPrefix(dir, stopAt) {
		if err := os.Remove(dir); err != nil {
			break
		}
		dir = filepath.Dir(dir)
	}
}
